libvast/src/option_declaration_set.cpp

```diff
@@ -0,0 +1,296 @@
+option_declaration_set::parse(option_map& xs, argument_iterator begin,
+                              argument_iterator end) const {
+  // Add all default values to the map.
+  for (auto& [long_name, x] : long_opts_) {
+    if (auto res = xs.add(long_name, x->default_value()); !res)
+      if (long_name != "help")
+        return std::make_pair(parse_state::option_already_exists, end);
+  }
+  auto parse_argument = [&](auto idx, const auto& option, auto begin, auto end) {
+    auto make_result = [](auto state, auto it, auto result) {
+      return std::make_pair(std::make_pair(state, it), std::move(result));
+    };
+    if (begin == end)
+      return make_result(parse_state::arg_declared_but_not_passed, begin,
+                         data{});
+    auto arg = begin->substr(idx);
+    auto [state, result] = option->parse(arg);
+    if (state != parse_state::successful)
+      return make_result(state, begin, data{});
+    return make_result(state, ++begin, std::move(result));
+  };
+  auto parse_short_option = [&](auto begin, auto end) {
+    // Extract the short name one of the fllowing strings: 
+    // "-s", "-sXX", ["-s" "XX"]
+    auto& x = *begin;
+    auto indicator = 1u; // char count of "-"
+    if (x.size() <= indicator)
+      return std::make_pair(parse_state::name_not_declartion, begin);
+    auto short_name = x[1];
+    // Parse the argument if available.
+    auto it = short_opts_.find(short_name);
+    if (it == short_opts_.end())
+      return std::make_pair(parse_state::name_not_declartion, begin);
+    auto& [_, option] = *it;
+    auto& long_name = option->long_name();
+    // Parse the argument if available.
+    if (option->has_argument()) {
+      std::pair<parse_state, decltype(begin)> res;
+      data argument;
+      if (x.size() > indicator + 1) {
+        std::tie(res, argument)
+          = parse_argument(indicator + 1, option, begin, end);
+      } else {
+        std::tie(res, argument) = parse_argument(0, option, ++begin, end);
+      }
+      if (res.first != parse_state::successful)
+        return res;
+      xs.set(long_name, argument);
+      return std::make_pair(parse_state::in_progress, res.second);
+    } else {
+      if (x.size() > indicator + 1)
+        return std::make_pair(parse_state::arg_passed_but_not_declared, begin);
+      xs.set(long_name, true);
+      return std::make_pair(parse_state::in_progress, ++begin);
+    }
+  };
+  auto parse_long_option = [&](auto begin, auto end) {
+    // Extract the long_name from one of the following strings: 
+    // "--long_name", "--long_name=XX".
+    auto& x = *begin;
+    auto idx = x.find('=');
+    auto indicator = 2u; // char count of "--"
+    auto long_name = x.substr(indicator, idx - indicator);
+    // Searches for the releated option.
+    auto it = long_opts_.find(long_name);
+    if (it == long_opts_.end())
+      return std::make_pair(parse_state::name_not_declartion, begin);
+    auto& [_, option] = *it;
```

Comment:
Same here, I dislike polluting the scope with unused variables only to avoid `it->second`.